### PR TITLE
Add chapter 16

### PR DESCRIPTION
**Pull request by OrangeToucan**

### CODE REVIEW

File: docs/16-errors-frames-etc.markdown [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+new `FRAME` -- see below). This `SUBR` drops into an infinite
+`READ-EVAL-PRINT` loop, which can be left via `ERRET` (section 16.4).
+
+The standard `LISTEN` loop has two features for getting a handle on
+objects that you have typed in and MDL has typed out. If the `ATOM`
+`L-INS` has a local value that is a `LIST`, `LISTEN` will keep recent
+inputs (what `READ` returns) in it, most recent first. Similarly, if
+the `ATOM` `L-OUTS` has a local value that is a `LIST`, `LISTEN` will
+keep recent outputs (what `EVAL` returns) in it, most recent first.
+The keeping is done before the `PRINT`ing, so that ^S does not defeat
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: kbd tag <kbd>^S</kbd>

File: docs/16-errors-frames-etc.markdown [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+    LISTENING-AT-LEVEL i PROCESS p
+
+where *i* is an integer (`FIX`) which is incremented each time
+`LISTEN` is called recursively, and *p* is an integer identifying the
+`PROCESS` (chapter 20) in which the `LISTEN` was `EVAL`ed. `LISTEN`
+then does `<APPLY <VALUE REP>>`, if there is one, and if it is
+`APPLICABLE`. If not, it applies the `SUBR` `REP` (without making a
+new `FRAME` -- see below). This `SUBR` drops into an infinite
+`READ-EVAL-PRINT` loop, which can be left via `ERRET` (section 16.4).
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: separate backticks for `READ`, `EVAL`, `PRINT`

File: docs/16-errors-frames-etc.markdown [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+(.L-INS NEWEST NEWER)
+<SET FOO 69>$
+69
+<SET FIXIT <2 .LINS>>	;"grab the last input"$
+<SET FOO 69>
+.L-INS$
+(.L-INS <SET FIXIT <2 .L-INS>> <SET FOO 69>)
+<PUT .FIXIT 3 105>$
+<SET FOO 105>
+<EVAl .FIXIT>$
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: uppercase EVAL

File: docs/16-errors-frames-etc.markdown [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+    <ARGS frame>
+
+("arguments") returns the argument `TUPLE` of *frame*.
+
+16.3.2. FUNCT
+
+    <FUNCT frame>
+
+("function} returns the `ATOM` whose `G/LVAL` is being applied in
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: punctuation around ("function")

no backticks around G/LVAL (oddly enough)

File: docs/16-errors-frames-etc.markdown [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+[a 1]
+<ERRET 5>$	;"This causes the + to return 5."
+15		;"finally returned by the *"
+
+
+Note that when you are in a call to `ERROR`, the most recent set of
+bindings is still in effect. This means that you can examine values of
+dummy variables while still in the error state. For example,
+
+<DEFINE F (A "AUX" (B "a string"))
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: begin code block

File: docs/16-errors-frames-etc.markdown [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+	#DECL ((VALUE) LIST (A) STRUCTURED (B) STRING)
+	(.B <REST .A 2>)	;Return this LIST. >$
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: quotes around "Return this LIST."

File: docs/16-errors-frames-etc.markdown [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+F
+<F '(1)>$
+
+*ERROR
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: asterisk after ERROR

File: docs/16-errors-frames-etc.markdown [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+<* 3 <+ a 1>>$
+*ERROR*
+ARG-WRONG-TYPE
++
+LISTENING-AT-LEVEL 2 PROCESS 1
+<ARGS <FRAME <FRAME>>>$
+[a 1]
+<ERRET 5>$	;"This causes the + to return 5."
+15		;"finally returned by the *"
+
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: end code block

File: docs/16-errors-frames-etc.markdown [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+    <RETRY frame>
+
+causes the control stack to be stripped down just beyond *frame*, and
+then causes the Subroutine call that generated *frame* to be done
+again. *frame* is optional, by default the `FRAME` of the last
+invocation of `ERROR` or `LISTEN`. `RETRY` differs from `AGAIN` in
+that (1) it is not intended to be used in programs; (2) it can retry
+any old *frame* (any Subroutine call), whereas `AGAIN` requires an
+`ACTIVATION` (`PROG` or `REPEAT` or "`ACT`"); and (3) if it retries
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: backticks should be outside the quotes in `"ACT"` (it's a string keyword in a function arg spec)

File: docs/16-errors-frames-etc.markdown [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+the `EVAL` of a `FORM` that makes an `ACTIVATION`, it will cause
+rebinding in the argument `LIST`, thus duplicating side effects.
+
+16.6. UNWIND
+
+`UNWIND` is an `FSUBR` that takes two arguments. usually `FORM`s. It
+`EVAL`s the first one, and, if the EVAL returns normally, the value of
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: backticks around the second `EVAL`

File: docs/16-errors-frames-etc.markdown [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+the `EVAL` call is the value of `UNWIND`. If, however, during the
+`EVAL` a non-local return attempts to return below the `UNWIND FRAME`
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: separate backticks around UNWIND and FRAME (it's the `FRAME` created by the call to `UNWIND`)

File: docs/16-errors-frames-etc.markdown [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+<DEFINE CLEAN ACT ("AUX" (C <OPEN "READ" "A FILE">))
+	#DECL ((C) <OR CHANNEL FALSE> ...)
+	<COND (.C
+		<UNWIND <PROG () ... <CLOSE .C>>
+			<CLOSE .C>>)>>
+```
+
+16.7. Control-G (^G)
+
+Typing control-G (^G, <ASCII 7>) at MDL causes it to act just as if an
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: kbd tag around <kbd>^G</kbd>, backticks around `<ASCII 7>`

File: docs/16-errors-frames-etc.markdown [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+error had occurred in whatever was currently being done. You can then
+examine the values of variables as above, continue by applying `ERRET`
+to one argument (which is ignored), `RETRY` a `FRAME` lower on the
+control stack, or flush everything by applying `ERRET` to no
+arguments.
+
+16.8. Control-S (^S)
+
+Typing control-S (^S, <ASCII 19>) at MDL causes it to stop what is
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: kbd tag around <kbd>^S</kbd>, backticks around `<ASCII 19>`

File: docs/16-errors-frames-etc.markdown [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+happening and return to the `FRAME` `.LERR\ !-INTERRUPTS`, returning
+the `ATOM` `T`. (In the Texex and Tops-20 versions, ^O also has the
```
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: Texex -> Tenex

kbd tag around <kbd>^O</kbd>

File: docs/16-errors-frames-etc.markdown (hunk 10 above)
Comment (taradinoc) [inline docs/16-errors-frames-etc.markdown]: also, period -> comma before "usually"

**Comment by OrangeToucan** [comment]
These changes have been made. I have added the keyboard tag although I should point out that it's non-standard: Only GitHub renders it. If someone were render the markdown any other way (say by using pandoc to convert the markdown to HTML) it would not be shown. (This is why I like to call the markdown used here as "GitHub's markdown", because it's their own special flavor of it.)

**Comment by taradinoc** [comment]
Looks like the code block in 16.4 still needs to be trimmed - the paragraph beginning "Note that when you are in a call to `ERROR`" should not be part of the code block.

**Comment by OrangeToucan** [comment]
Oh yes, I see that. This has been changed.

**Pull request merged**